x-pack/legacy/plugins/code/server/distributed/cluster/cluster_service.ts

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import * as lodash from 'lodash';
+import { ClusterMeta } from './cluster_meta';
+import { EsClient } from '../../lib/esqueue';
+import { RepositoryObjectClient } from '../../search';
+import { Poller } from '../../poller';
+import { Logger } from '../../log';
+import { Repository } from '../../../model';
+import { RoutingTable } from './routing_table';
+import { ClusterState } from './cluster_state';
+import { ClusterStateEvent } from './cluster_state_event';
+
+/**
+ * ClusterService synchronizes the core cluster states with the remote, and provides methods to read the current state,
+ * and to register listeners to watch the state change.
+ */
+export class ClusterService {
+  private readonly clusterStateListeners: ClusterStateListener[] = [];
+  private readonly clusterStatePoller = new Poller<void>({
+    functionToPoll: () => {
+      return this.pollClusterState();
+    },
+    pollFrequencyInMillis: 5000,
```

Comment:
nit: have a const var for this `5000`

Reply:
If we go with this mechanism, should we make it a configurable property?

Reply:
probably no for now.